Library/OcAppleKernelLib/CachelessContext.c

```diff
@@ -0,0 +1,1142 @@
+    }
+    FreePool (DependKext);
+  }
+
+  FreePool (BuiltinKext);
+}
+
+STATIC
+BOOLEAN
+AddKextDependencies (
+  IN OUT LIST_ENTRY           *Dependencies,
+  IN     XML_NODE             *InfoPlistLibraries
+  )
+{
+  DEPEND_KEXT       *DependKext;
+  UINT32            ChildCount;
+  UINT32            ChildIndex;
+  CONST CHAR8       *ChildPlistKey;
+
+  LIST_ENTRY        *KextLink;
+  BOOLEAN           DependencyExists;
+
+  ChildCount = PlistDictChildren (InfoPlistLibraries);
+  for (ChildIndex = 0; ChildIndex < ChildCount; ChildIndex++) {
+    ChildPlistKey = PlistKeyValue (PlistDictChild (InfoPlistLibraries, ChildIndex, NULL));
+    if (ChildPlistKey == NULL) {
+      continue;
+    }
+
+    DependencyExists  = FALSE;
+    KextLink          = GetFirstNode (Dependencies);
+    while (!IsNull (Dependencies, KextLink)) {
+      DependKext = GET_DEPEND_KEXT_FROM_LINK (KextLink);
+
+      if (AsciiStrCmp (DependKext->BundleId, ChildPlistKey) == 0) {
+        DependencyExists = TRUE;
+        break;
+      }
+
+      KextLink = GetNextNode (Dependencies, KextLink);
+    }
+
+    if (!DependencyExists) {
+      DependKext = AllocateZeroPool (sizeof (*DependKext));
+      if (DependKext == NULL) {
+        return FALSE;
+      }
+      DependKext->Signature = DEPEND_KEXT_SIGNATURE;
+      DependKext->BundleId = AllocateCopyPool (AsciiStrSize (ChildPlistKey), ChildPlistKey);
+
+      InsertTailList (Dependencies, &DependKext->Link);
+    }
+  }
+
+  return TRUE;
+}
+
+STATIC
+EFI_STATUS
+ScanExtensions (
+  IN OUT CACHELESS_CONTEXT    *Context,
+  IN     EFI_FILE_PROTOCOL    *File,
+  IN     CONST CHAR16         *FilePath,
+  IN     BOOLEAN              ReadPlugins
+  )
+{
+  EFI_STATUS          Status;
+  EFI_FILE_PROTOCOL   *FileKext;
+  EFI_FILE_PROTOCOL   *FilePlist;
+  EFI_FILE_PROTOCOL   *FilePlugins;
+  EFI_FILE_INFO       *FileInfo;
+  UINTN               FileInfoSize;
+  BOOLEAN             UseContents;
+
+  CHAR8               *InfoPlist;
+  UINT32              InfoPlistSize;
+  XML_DOCUMENT        *InfoPlistDocument;
+  XML_NODE            *InfoPlistRoot;
+  XML_NODE            *InfoPlistValue;
+  XML_NODE            *InfoPlistLibraries;
+  CONST CHAR8         *TmpKeyValue;
+  UINT32              FieldCount;
+  UINT32              FieldIndex;
+
+  BUILTIN_KEXT        *BuiltinKext;
+  CHAR16              TmpPath[256];
+
+  DEBUG ((DEBUG_INFO, "OCAK: Scanning %s...\n", FilePath));
+
+  FileInfo = AllocatePool (SIZE_1KB);
+  if (FileInfo == NULL) {
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  File->SetPosition (File, 0);
+
+  do {
+    FileInfoSize = SIZE_1KB - sizeof (CHAR16);
+    Status = File->Read (File, &FileInfoSize, FileInfo);
+    if (Status == EFI_BUFFER_TOO_SMALL) {
+      FileInfoSize = SIZE_2KB - sizeof (CHAR16);
+      FileInfo = ReallocatePool (SIZE_2KB, SIZE_1KB, FileInfo);
+      Status = File->Read (File, &FileInfoSize, FileInfo);
+      //
+      // TODO? would the buffer ever be this big realistically.
```

Comment:
This should contain file info for a single file info, correct? I believe it is safe to assume it will not be that large in a valid scenario. For an invalid scenario (e.g. FS corruption) we should make sure that the failure here is handled and error and exits the function. Please add an error check and an explanation.